### ISSUE TITLE
Buttons with charges are greying out even when charges are still available

### ISSUE THREAD

**Issue by raysmith59** [brittyazel/Neuron#239]
Something like Ironskin Brew has 3 charges, as soon as one is consumed the button greys out which makes me think it isn't available.  I don't think it did that before and it's definitely not how I'd expect it to work.  Related to the Cooldown Transparency toggle in the bar config, I only want it to be active when there aren't any charges available.

**Comment by dpwrussell** [comment]
I can confirm this. It seems to be greyed out when it is `n - 1` charges rather than at zero available.

**Comment by AlexFolland** [comment]
This looks like it's working exactly as it always has to me.  It's showing the cooldown for the next charge.  I'm pretty sure nothing's changed there.

**Comment by dpwrussell** [comment]
I'm not using CT.

Here is how Neuron used to look when some charges were on CD, but some were available:

![wowscrnshot_010119_234715](https://user-images.githubusercontent.com/1889043/50577429-c269cf00-0e1f-11e9-89a7-28ea069ce5b1.jpg)

Notice how `2`, Judgement is on CD for one charge. The button (once GCD sweep has completed) is not greyed out at all, but the yellow sweep for the overall CD is subtly displayed. This is versus the screenshot I showed above where the button is greyed out for the entirety of the remaining CD of the next available charge.

**Comment by brittyazel** [comment]
Yep I can fix all of this. I rewrote the whole cooldown system, and this was an oversight on my part. Give me a little bit to get checks for these charge conditions implemented.

Also the other option in the GUI was for aura countdowns, and I fully broke it. I'm going to fix it I just haven't had a chance

**Comment by dpwrussell** [comment]
@brittyazel Hey, I just installed 1.0.3b and it doesn't seem to be fixed.

**Comment by TruthNZ** [comment]
Updated to 1.03d, I've still got the same issue.

**Comment by brittyazel** [comment]
What ability is the one doing this? I cannot make it happen with any of my abilities with charges and cooldowns

**Comment by dpwrussell** [comment]
@brittyazel It happens for me with any ability that has multiple charges, e.g. Judgement or Shield of the Righteous.

**Comment by TruthNZ** [comment]
Blood Death Knight: Blood Boil and Rune Strike

**Comment by raysmith59** [comment]
Based on my testing the only issue I'm seeing with charges at this point is that when an ability has more than one charge, all are used up (causing the button to go grey) and one charge accrues the button isn't reacting the way it used to.

Before:
![image](https://user-images.githubusercontent.com/2917594/50743281-d036a000-11e3-11e9-93d5-931964e3b49c.png)

Now:
![image](https://user-images.githubusercontent.com/2917594/50743291-f9efc700-11e3-11e9-9689-054ac23f88ed.png)

The button does come back as the sweep progresses, though, so I think the issue is just that the button is too grey when it's pre-swept where before the grey was very subtle or there wasn't any darkening at all and the only function of the sweep was to show the progress of the cooldown, not to impact the look of the button.

**Comment by brittyazel** [comment]
Hmm...I'm not sure what to do with this. The reason it is "dark" is like you said, it's showing the cooldown spiral and it's just at the apex point. This is very different than the cdAlpha issue that was the original issue. In this case the button is doing exactly as I programmed it to. I look into the blizzard codebase and see hwo they handle it. I think there is a charge cooldown function that does the cooldown on the edge of the button only. Maybe we can use that

So we don't have any control on how the Facebook spiral works. It's a set color, speed, etc.

**Comment by raysmith59** [comment]
Well, if you want to use the Blizzard default frames for reference they're closer to what we had before than what we have now.

No charges remaining:
![image](https://user-images.githubusercontent.com/2917594/50743955-b9944700-11eb-11e9-88b1-a69d9f7fb879.png)

First charge almost but not quite refreshed:
![image](https://user-images.githubusercontent.com/2917594/50743968-d761ac00-11eb-11e9-97d1-a00256d2c656.png)

One charge refreshed, 2nd still on CD:
![image](https://user-images.githubusercontent.com/2917594/50743977-f3fde400-11eb-11e9-9ccb-f07a10012b99.png)

Basically the button looks the same when you use a charge as when a charge refreshes which makes sense.  Currently in Neuron that isn't the case, the charge refresh should look the same as the charge used (charge used looks right, IMO).  If you can duplicate the Blizzard behaviour I think it'll satisfy everyone here.

**Comment by brittyazel** [comment]
Alright, I think I fixed it. Can you test the version in MASTER for me to see if this addressed both your issues? I believe I have recreated how the blizzard default frames handle cooldowns with charges now. Please put it through the ringer if you don't mind, and let me know you believe it's ready for me to release an update to everyone

**Comment by dpwrussell** [comment]
I tried it from master and it looks like it's getting towards a fix. The GCD is not being displayed properly though. Top bars are Neuron, bottom are default blizzard UI.

Here you can see Sheild of the Righteous (off the regular GCD, but obviously has its own personal GCD.

![wowscrnshot_010719_185828](https://user-images.githubusercontent.com/1889043/50787450-5fa88580-12ae-11e9-98aa-4986f13053e1.jpg)


And here is Judgement, which is on the regular GCD. All the other GCD abilities correctly display the GCD, but not the one that was used.

![wowscrnshot_010719_185356](https://user-images.githubusercontent.com/1889043/50787364-1bb58080-12ae-11e9-86f4-6822a78614eb.jpg)

Appreciate your efforts on this @brittyazel

**Comment by raysmith59** [comment]
Aside from the issue dpw mentions above (to maybe clarify since I wasn't quite sure until I tested it, the GCD sweep is working fine for buttons that don't have charges but ones that do have charges don't show the GCD sweep when they're the button that's pushed), the only issue I see that remains is that while the GCD sweep re-activates the button in the area where the sweep has happened (and is greyed out in the area where the sweep hasn't happened, you can see it in the last screenshot above with the 3 buttons at the bottom), the same isn't happening for buttons with cooldowns, in the case of cooldowns the button remains grey until the sweep completes as can be seen here:

![image](https://user-images.githubusercontent.com/2917594/50793076-02540a80-1295-11e9-8d09-287a548002f5.png)

The sweep is at 9:00 in that screenshot so 3/4 of the button should have re-activated at that point.

On the upside, the timing of the buttons no longer feels laggy and seems to be working as it did before.  We're close.

**Comment by TruthNZ** [comment]
Much better as of now for me, running the master branch.
Although as a side note: keybindings aren't working on my non-main action bar for some strange reason (they were in 1.0.3d).

**Comment by dpwrussell** [comment]
Sorry, but it's still not fixed I'm afraid.

* Two charges of Crusader Strike are available.
* I press Crusader Strike
* Expected result is that CS shows the GCD, followed by available.
* Actual result is that CS shows as if available instantly. You can see the GCD being displayed on all my other abilities, while CS is showing available. If you press it, it is not available.

![wowscrnshot_011619_211132](https://user-images.githubusercontent.com/1889043/51278873-a21b4200-19d3-11e9-8a96-14d52163bdd6.jpg)

**Comment by brittyazel** [comment]
Interesting, I'll look at this.

**Comment by brittyazel** [comment]
Nvm it seems stable enough for me to push a release. 1.0.5 coming at you in about 30min

**Comment by dpwrussell** [comment]
@brittyazel Looks great! Thanks!

**Comment by brittyazel** [comment]
I also added three cooldown distinction levels. It's subtle but I think it's nice. 
1) Regular spell cooldowns have the dark sweeping area effect + the leading edge has a yellow highlight
2) Charge cooldowns that still have available charges have JUST the yellow leading edge (not the darkend sweeping)
3)GCD cooldowns have just the dark sweeping area effect and not the yellow leading edge

This adds a subtle, but perceptible difference to the different CD types. Maybe this can help people when they're playing

**Comment by dpwrussell** [comment]
I haven't noticed that yet, but I'll have a look. The reduced latency is amazing. It was so painful before. Massive improvement!!

**Comment by dpwrussell** [comment]
One more related problem I'm afraid. I'm not sure what it is about this ability that makes it not behave 
like the other multi-stack ones. You can see that Blessed Hammer (between judgement and consecration) has 0 stacks, but it showing as available.

![wowscrnshot_012119_085124](https://user-images.githubusercontent.com/1889043/51463049-236b2f80-1d5a-11e9-829b-8aa63f2b8242.jpg)
